### ISSUE TITLE
pgen 'disk' behaving strangely on KNL with --eos adiabatic

### ISSUE THREAD

**Issue by ncimerman** [PrincetonUniversity/athena#435]
I am currently experiencing strange behaviour of the code for using the pgen 'disk' when running on KNL. More specifically, configuring with

    python3 configure.py --prob disk --eos isothermal --coord cylindrical --flux hlle --nghost=2 --cxx icpc-phi -mpi -omp

produces code that compiles and runs fine with the standard input file

    -i inputs/hydro/athinput.disk_cyl

When switching to an adiabatic EoS

    python3 configure.py --prob disk --eos adiabatic --coord cylindrical --flux hlle --nghost=2 --cxx icpc-phi -mpi -omp

the timestep plummets to zero after four time-steps or so.

This issue does not occur when disabling OpenMP SIMD (adding --cflag='-qno-openmp-simd'). A closer look at the output data at t=1e-4 revealed that within one time-step there is a characteristic pattern of erroneous values, for example in density (see attached).

After first spending a lot of time going through changes in my version of the code, I have managed to reproduce this using the current version of this public repository.

I have tested Intel compiler versions 2018.3 and 2019.3 with both behaving the same. I also tested two Riemann solvers: hlle and roe, both giving the same.

Testing the 'blast' pgen in cylindrical coordinates (which is 3D however), it appears that this pgen does not show this behaviour.

In summary, this seems to be an issue related to OpenMP SIMD and depends on the equation of state (isothermal vs adiabatic) and might be specific to 2D setups. I would be very grateful if someone could try and reproduce this, perhaps also with other compilers. I am running out of ideas and getting worried that I made some silly mistake somewhere.
![lin_rho_iso_f](https://user-images.githubusercontent.com/51494537/162909544-3a75e6ad-6c66-4d39-802e-6c2ca84cd874.png)

**Comment by ncimerman** [comment]
After writing this, I investigated what happens when the disk problem is run in 3D by setting

<mesh>
nx3         = 1

<orbital_advection>
OAorder     = 2

in the input file.

And indeed, the time-step does not drop to zero immediately!

This might hint at some 3D loops being optimized in an unsafe or wrong way when the setup is 2D. Switching orbital advection on/off does not make a difference to the issue in 2D.

**Comment by c-white** [comment]
Can you report what happens if you run with pure MPI (remove `-omp` from the configure step)? This will test if there is a thread-safety issue.

**Comment by ncimerman** [comment]
Just to double-check: Removing `-omp` gives makes my configure step

`python3 configure.py --prob disk --eos adiabatic --coord cylindrical --flux hlle --nghost=2 --cxx icpc-phi -mpi`

The `-qopenmp` flag disappears, but I still see the `-qopenmp-simd` flag during compilation, which is switched on due to `--cxx icpc-phi` if I understand correctly from looking configure.py.

Findings:

1) I tested only removing `-omp` and it still produces the error. 

2) Removing `-omp` **and** adding `-qno-openmp-simd` however works fine. It does not make sense to me why the latter should make a difference if OpenMP is in fact deactivated.

Researching this, I did find a thread that seemed to report a bug in ifort (2019) that caused OMP directives to be parsed, even when `-qno-openmp` was passed (if I understood correctly). See here: https://community.intel.com/t5/Intel-Fortran-Compiler/ifort-2019-parsing-omp-directives-with-qno-openmp/td-p/1142889

**Comment by felker** [comment]
It is confusingly worded, but OpenMP SIMD pragmas and OpenMP threading are controlled by separate compiler options. So this seems to be an issue with one or several `#pragma omp simd` uses, not anything related to threading. You could try selectively commenting out some of those to find the problematic ones.

**Comment by ncimerman** [comment]
I see, I had indeed understood those flags differently, thanks for clarifying this. Given that there are no `#pragma omp simd` in the disk pgen, the issue seems to be related to another part of the code.

I just did another thing that I should have done much sooner: looking at the initial data in both primitive and conserved quantities. The primitive values, which are set in the pgen and which I had only looked at so far, are fine. However, the conserved variables show the same error pattern from t=0.

So I had a look at `EquationOfState::PrimitiveToConserved` in `src/eos/adiabatic_hydro.cpp` and found that there is a pragma different to the usual ones, and for example in `void EquationOfState::ConservedToPrimitive`. What is the reason that the pragmas are different in the two functions?

I commented out the `#pragma omp simd` in line 96 and indeed that fixed the issue!

**Comment by ncimerman** [comment]
I also just checked and found that in `src/eos/isothermal_hydro.cpp` both functions have the same loop structure in terms of pragmas in contrast to `src/eos/adiabatic_hydro.cpp`

**Comment by ncimerman** [comment]
I have just tested giving `ConservedToPrimitive` the same loop structure as `PrimitiveToConserved` (i.e. only one `#pragma omp simd` inside k-,j- and outside i-loop) and can confirm that it seems to work fine in my case as well.

**Comment by c-white** [comment]
That pragma does look out of place. @felker It seems like it was introduced in https://github.com/PrincetonUniversity/athena/commit/e33f62e4d73622073447ba5413ca2bc35f654329. If this really is a bug, though, it's weird that it hasn't caused any issues for over 4 years.

**Comment by felker** [comment]
`#pragma novector` is an Intel-specific compiler directive; that might explain why problems are rare (plus it only seems to affect 2D cylindrical coordinates pgens?). The pragma should be legal... maybe a compiler bug. But I am ok to revert it to the standard inner-loop SIMD pragma. 

@ncimerman what happens if you comment out that pragma, leaving `#pragma omp simd` active on the outer loop?

**Comment by ncimerman** [comment]
Just tested leaving only the `#pragma omp simd`
```
#pragma omp simd                                                                                                                                             
  for (int k=kl; k<=ku; ++k) {                                                                                                                               
    for (int j=jl; j<=ju; ++j) {                                                                                                                             
//       #pragma omp simd                                                                                                                                     
//#pragma novector                                                                                                                                           
      for (int i=il; i<=iu; ++i) {
```
and it produces the error as well. So `#pragma novector` does not seem to be the culprit.

I agree, this only seems to happen in 2D adiabatic simulations, I haven't tested if the coordinates make a difference, but I can imagine it affecting other coordinate systems as well.

So the remedy seems to be to revert to the inner-loop SIMD.
